network/ClientNetworking.cpp

```diff
@@ -283,94 +332,84 @@ void ClientNetworking::SendSynchronousMessage(Message message, Message& response
 }
 
 void ClientNetworking::HandleConnection(tcp::resolver::iterator* it,
-                                        boost::asio::deadline_timer* timer,
                                         const boost::system::error_code& error)
 {
     if (error) {
-        if (!m_cancel_retries) {
-            if (TRACE_EXECUTION)
-                DebugLogger() << "ClientNetworking::HandleConnection : connection error ... retrying";
-            m_socket.async_connect(**it, boost::bind(&ClientNetworking::HandleConnection, this,
-                                                     it,
-                                                     timer,
-                                                     boost::asio::placeholders::error));
-        }
+        if (TRACE_EXECUTION)
+            DebugLogger() << "ClientNetworking::HandleConnection : connection "
+                          << "error #"<<error.value()<<" \"" << error.message() << "\""
+                          << "... retrying";
     } else {
         if (TRACE_EXECUTION)
             DebugLogger() << "ClientNetworking::HandleConnection : connected";
-        timer->cancel();
+
         boost::mutex::scoped_lock lock(m_mutex);
-        m_connected = true;
+        m_rx_connected = true;
+        m_tx_connected = true;
     }
 }
 
-void ClientNetworking::CancelRetries()
-{ m_cancel_retries = true; }
-
 void ClientNetworking::HandleException(const boost::system::system_error& error) {
-    if (error.code() == boost::asio::error::eof ||
-        error.code() == boost::asio::error::connection_reset ||
-        error.code() == boost::asio::error::operation_aborted)
-    {
-        DebugLogger() << "ClientNetworking::NetworkingThread() : Networking thread will be terminated "
-                      << "due to disconnect exception \"" << error.what() << "\"";
-    } else {
+    if (error.code() == boost::asio::error::eof) {
+        DebugLogger() << "Client connection disconnected by EOF from server.";
+        m_socket.close();
+    }
+    else if (error.code() == boost::asio::error::connection_reset)
+        DebugLogger() << "Client connection disconnected, due to connection reset from server.";
+    else if (error.code() == boost::asio::error::operation_aborted)
+        DebugLogger() << "Client connection closed by client.";
+    else {
         ErrorLogger() << "ClientNetworking::NetworkingThread() : Networking thread will be terminated "
-                      << "due to unhandled exception \"" << error.what() << "\"";
+                      << "due to unhandled exception error #" << error.code().value() << " \""
+                      << error.code().message() << "\"";
     }
 }
 
-void ClientNetworking::NetworkingThread() {
+void ClientNetworking::NetworkingThread(std::shared_ptr<ClientNetworking>& self) {
+    auto protect_from_destruction_in_other_thread = self;
     try {
```

Comment:
no need to check, after making the copy, that self has been destructed in another thread between the call and the copy was made?